pkg/assets/containers.go

```diff
@@ -0,0 +1,325 @@
+
+	}
+
+	asset, err := GetContainerRegistry(clusterSpec)
+
+	if err != nil {
+		return "", fmt.Errorf("unable to get container registry, asset: %v", err)
+	}
+
+	// registry is not set
+	if asset == nil {
+		return "", nil
+	}
+
+	return getRegistry(asset), nil
+
+}
+
+func getRegistry(asset *kops.ContainerAsset) string {
+
+	if asset == nil {
+		return ""
+	}
+
+	buf := new(bytes.Buffer)
+
+	if asset.Domain != "" {
+		buf.WriteString(asset.Domain)
+		buf.WriteString("/")
+	}
+
+	if asset.Name != "" {
+		buf.WriteString(asset.Name)
+	}
+
+	return buf.String()
+}
+
+func getRepoContainer(registry string, asset *kops.ContainerAsset) (string, error) {
+
+	if asset == nil {
+		return "", fmt.Errorf("asset cannot be nil")
+	}
+	buf := new(bytes.Buffer)
+
+	if registry != "" {
+		buf.WriteString(registry)
+		buf.WriteString("/")
+	} else if asset.Domain != "" {
+		buf.WriteString(asset.Domain)
+		buf.WriteString("/")
+	}
+
+	if asset.Name != "" {
+		buf.WriteString(asset.Name)
+	}
+
+	if asset.Tag != "" {
+		buf.WriteString(":")
+		buf.WriteString(asset.Tag)
+	}
+
+	if asset.Digest != "" {
+		buf.WriteString("@")
+		buf.WriteString(asset.Digest)
+	}
+
+	return buf.String(), nil
+}
+
+// GetContainerRegistry returns a ContainerAsset get the asset container registry if a cluster has an registry.
+func GetContainerRegistry(clusterSpec *kops.ClusterSpec) (*kops.ContainerAsset, error) {
+
+	if clusterSpec == nil {
+		return nil, fmt.Errorf("unable to parse assets container registry as cluster spec is nil")
+
+	}
+
+	if clusterSpec.Assets != nil && clusterSpec.Assets.ContainerRegistry != nil {
+		registry := strings.TrimSuffix(*clusterSpec.Assets.ContainerRegistry, "/")
+		asset, err := ParseContainer(registry)
+
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse assets container registry api value: %v", registry)
+		}
+
+		return asset, nil
+	}
+
+	return nil, nil
+}
+
+// GetContainerAndRegistryAsString returns a full container string if a cluster has an asset container registry.
+func GetContainerAndRegistryAsString(clusterSpec *kops.ClusterSpec, container string) (string, error) {
+
+	if clusterSpec == nil {
+		return "", fmt.Errorf("unable to parse assets container registry as cluster spec is nil")
+
+	}
+
+	registry, err := GetContainerRegistry(clusterSpec)
+
+	if err != nil {
+		return "", err
+	}
+
+	r := getRegistry(registry)
+
+	asset, err := ParseContainer(container)
+
+	if err != nil {
+		return "", err
+	}
+
+	container, err = getRepoContainer(r, asset)
+
+	if err != nil {
+		return "", err
+	}
+
+	return container, nil
+}
+
+// GetContainerAsString returns a full parsed container string.
+func GetContainerAsString(container string) (string, error) {
+	asset, err := ParseContainer(container)
+
+	if err != nil {
+		return "", fmt.Errorf("unable to parse container: %+v: %v", container, err)
+	}
+
+	container, err = getRepoContainer("", asset)
+
```

Comment:
Is there a reason for the blank line in between the function which sets err and checks it?  I think it's nice to keep them together...